Userland/Utilities/cat.cpp

```diff
@@ -23,39 +21,34 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
     args_parser.add_positional_argument(paths, "File path", "path", Core::ArgsParser::Required::No);
     args_parser.parse(arguments);
 
-    Vector<int> fds;
+    Vector<NonnullRefPtr<Core::File>> files;
+    TRY(files.try_ensure_capacity(paths.size()));
     if (paths.is_empty()) {
-        TRY(fds.try_append(STDIN_FILENO));
+        TRY(files.try_append(Core::File::standard_input()));
     } else {
         for (auto const& path : paths) {
-            int fd;
             if (path == "-") {
-                fd = 0;
+                files.unchecked_append(Core::File::standard_input());
             } else {
-                auto fd_or_error = Core::System::open(path, O_RDONLY);
-                if (fd_or_error.is_error()) {
-                    warnln("Failed to open {}: {}", path, fd_or_error.error());
-                    continue;
-                }
-                fd = fd_or_error.release_value();
+                auto file_or_error = Core::File::open(path, Core::OpenMode::ReadOnly);
+                if (file_or_error.is_error())
+                    warnln("Failed to open {}: {}", path, file_or_error.error());
+                else
+                    files.unchecked_append(file_or_error.release_value());
             }
-            TRY(fds.try_append(fd));
         }
     }
 
     TRY(Core::System::pledge("stdio", nullptr));
 
-    for (auto& fd : fds) {
+    for (auto& file : files) {
+        auto file_stream = Core::InputFileStream(file);
         for (;;) {
-            char buffer[32768];
-            auto nread = TRY(Core::System::read(fd, buffer, sizeof(buffer)));
-            if (nread == 0)
+            auto buffer = TRY(file_stream.try_read(32768));
```

Comment:
This does have a drawback in that it allocates the output now, could we come up with something less allocation-happy?

Reply:
In master this uses span instead, but retains the stack array.